### DISCUSSION TITLE
[help] TGI on Lynx

### DISCUSSION THREAD

**Discussion by Fabrizio-Caruso** [cc65/cc65#1688] [Q&A]
@karrika Thanks a lot for your examples! I am trying to figure out to construct
```
#define freeze_BPP          3
#define freeze_WIDTH        7
#define freeze_HEIGHT       7
const unsigned char freeze[] = {
    0x03,0xB6,0x80,0x03,0xA0,0x00,0x03,0x14,
    0x00,0x03,0x00,0x00,0x01,0x03,0xA0,0x00,
    0x03,0x14,0x00,0x03,0x00,0x00,0x01,0x03,
    0xB4,0x00,0x03,0xA2,0x80,0x03,0xA0,0x00,
    0x01,0x03,0x16,0xD0,0x03,0xB4,0x00,0x03,
    0xA2,0x80,0x03,0xA0,0x00,0x00,
};
```
from my file where I have defined freeze (8x8) as:
```
0,60,66,207,195,102,36,24
```

It seems you use the external tool sp65. I need to do the conversion in a script. What is the structure of the bytes in the array.
You have 46 bytes, for 49 pixels with 3 bit per pixels (8 colors but the comments say 6 colors).
So it looks you not takin the whole 8x8 shape but you are cutting it as 7x7.

**Comment by karrika** [comment]
I just realized that the sp65 on this repo has not been updated since the pcx file format changed. So it is kind of hard to do without fixing the sp65 first.

However... I did a few versions so you can see what freeze looks like:
/*
 * This file was generated by sp65 2.13.9 from
 * Slice of tiles.pcx (7x7, 4 colors, indexed)
 */

#define freeze_BPP          2
#define freeze_WIDTH        7
#define freeze_HEIGHT       7
const unsigned char freeze[] = {
    0x03,0x54,0x00,
    0x03,0x40,0x00,
    0x03,0x10,0x00,
    0x03,0x00,0x00,0x01,
    0x03,0x40,0x00,
    0x03,0x10,0x00,
    0x03,0x00,0x00,0x01,
    0x03,0x50,0x00,
    0x03,0x44,0x00,
    0x03,0x40,0x00,0x01,
    0x03,0x15,0x00,
    0x03,0x50,0x00,
    0x03,0x44,0x00,
    0x03,0x40,0x00,
    0x00,
};

The file had an anchor point. This is where the drawing quadrant will change. It starts drawing from the center and draws one quadrant right and down. Then one quadrant up and right, One left and up. Final left and down.

**Comment by karrika** [comment]
Now I skip the anchor point and create a single colour sprite.

```
#define freeze_BPP          1
#define freeze_WIDTH        7
#define freeze_HEIGHT       7
const unsigned char freeze[] = {
    0x03,0x10,0x00,
    0x03,0x54,0x00,
    0x03,0x38,0x00,
    0x03,0xEE,0x00,
    0x03,0x38,0x00,
    0x03,0x54,0x00,
    0x03,0x10,0x00,
    0x00,
};
```

Can you recognize the shape?

**Reply by Fabrizio-Caruso** [reply]
I don't understand the representation TGi use. I see 22 bytes but I expected only 7 for a 7x7 monochromatic shape. I don't know what those values represent? Is it a standard bitmap representation with bits representing pixels? Is it a vectorial représentation? Which algorithm do I need to use to transform my simple bitmap representation to the TGI one?

**Reply by karrika** [reply]
> I don't understand the representation TGi use. I see 22 bytes but I expected only 7 for a 7x7 monochromatic shape. I don't know what those values represent? Is it a standard bitmap representation with bits representing pixels? Is it a vectorial représentation? Which algorithm do I need to use to transform my simple bitmap representation to the TGI one?

Every line I wrote represents a line of pixels. count + data + zero
```
0x10 = ...X....
0x54 = .X.X.X..
0x38 = ..XXX...
0xEE = XXX.XXX.
0x38 = ..XXX...
0x54 = .X.X.X..
0x10 = ...X....

```

**Reply by karrika** [reply]
If you use sprites you can colour the bitmap with a pen palette.
```
SCB_REHV_PAL Sfreeze = {
        BPP_1 | TYPE_NORMAL,
        LITERAL | REHV,
        NO_COLLIDE,
        0,
        freeze,
        hpos, vpos,
        0x100, 0x100,
        {0x04}
};

```
In this case the penpal has value 04. 0 = transparency, 4 = fourth color of your palette

**Reply by Fabrizio-Caruso** [reply]
Thanks! Count is always 3. I guess is the number of bytes per line including count and zero.
I guess a color is implemented as an additional bit plane.
So with 2 bit places you would have 2 lines of data to descrive a 4 color line,
Is this representation assuming the left most pixel is not set? There is a bug in the hardware. Does this reprentation works anyway if all pixels are set or do we have to keep the left most pixel unset?

**Reply by Fabrizio-Caruso** [reply]
So if I want no transparency, I probably need 2BPP as 0 bits are transparent. I would like to implement pseudo tiles (solid two color shapes). I could simply delete the area and then blit a sprite but it is probably inefficient.

**Reply by karrika** [reply]
The count can be anything. It depends on how many bytes the sprite engine has to work with. The pixels are 1,2,3 or 4 bits per pixel. The bits are packed as a stream in the bytes. Highest bit comes first.
The pixel value is then something between 0..15. To get the colour you use the pixel value as an index to the penpal at the end of the sprite. The value in penpal[pixel] is the colour index.

**Reply by karrika** [reply]
> So if I want no transparency, I probably need 2BPP as 0 bits are transparent. I would like to implement pseudo tiles (solid two color shapes). I could simply delete the area and then blit a sprite but it is probably inefficient.

No! You can have any two colours in a BPP_1 pixel

The penpal can have any numbers. 0x12 means that
0 becomes colour 1 and 
1 becomes colour 2

**Reply by Fabrizio-Caruso** [reply]
Is this representation assuming the left most pixel is never set? If I remember well tere is a bug in the hardware. Does this representation works even if all pixels are set or do we have to keep the left most pixel unset?

**Reply by karrika** [reply]
> Is this representation assuming the left most pixel is never set? If I remember well tere is a bug in the hardware. Does this representation works even if all pixels are set or do we have to keep the left most pixel unset?

No. You do not need to keep the left most pixel unset. The bug is only at the end of the line. If the last pixel of a sprite happens to be on the last bit of a byte the you need to add an extra zero byte. But this is only for packed sprites. Literal sprites have no bugs.

**Reply by karrika** [reply]
@jum on AtariAge created an online sprite editor for the Lynx. It is really useful.

Have a look https://github.com/james7780/tinyspriteeditor

Just download the code and run "firefox tinyspriteeditor_lynx.html". You can import your sprites in png format, edit them and output them as Lynx ASM or Lynx C files.

**Reply by Fabrizio-Caruso** [reply]
@karrika Thanks, I will look at it. It is handy for Lynx specific projects. It is not what I need for Cross-Lib because I need to generate target specific assets automatically from hardware agnostic assets for about 200 different targets:  https://github.com/Fabrizio-Caruso/CROSS-LIB

**Reply by Fabrizio-Caruso** [reply]
@karrika Sorry, but even literal sprites seem to show a bug (no red background right-most pixel on lines with even value) :
![image](https://user-images.githubusercontent.com/29057615/156853267-9f22d8cf-650f-4127-bba4-25c02ab82849.png)

I am trying to display a solid 8x8 block with two colors:
```
const unsigned char freeze[] = {
    0x02,170,
    0x02,255,
    0x02,170,
    0x02,255,
    0x02,170,
    0x02,255,
    0x02,170,
    0x02,255,
    0x00,
};
```
If I use 0x03,170,0 I get 8 extra solid background pixels. So there is no straightforward way to display just an 8x8 solid block.

**Reply by Fabrizio-Caruso** [reply]
I may assume that the background color of my block is the same as the general background, it looks correct and I may be able to delete it. So maybe I can get what I need somehow but I think the bug is present even with literal sprites.

**Comment by mrdudz** [comment]
> since the pcx file format changed

??? What changes are you referring to? PCX is ancient, isnt it?

**Reply by karrika** [reply]
> > since the pcx file format changed
> 
> ??? What changes are you referring to? PCX is ancient, isnt it?

Not exactly. The latest GIMP is using different techniques that compresses the format. sp65 cannot open these files anymore. I believe the change happened in 2020 or so. Anyway it was impossible for me to use the new version of Gimp so I had to fix it. My plan was always to bring the improvements also to the official cc65 but i just forgot...

**Reply by mrdudz** [reply]
I have serious doubts that the reason is "the PCX format changed" though - more likely is that sp65 never supported it 100% correctly and gimp just writes them out in a slightly different way now. I cant find any source that claims the RLE compression ever changed for that matter :)

**Reply by karrika** [reply]
> I have serious doubts that the reason is "the PCX format changed" though - more likely is that sp65 never supported it 100% correctly and gimp just writes them out in a slightly different way now. I cant find any source that claims the RLE compression ever changed for that matter :)

That is the most likely explanation. My faint memory says it had something to do with planes. But I also added other stuff like outputting palettes for the Lynx. And then I worked a long time with 42bs and laoo trying to squeeze more compression to the sprites. The old Lynx programmers saved much more bytes by hand than we were able to do with sprpck or sp65. So I am still in the middle of developing sp65.

**Reply by karrika** [reply]
Now I remember!. It had to do with pcx files with depth below 5 colours!
The way the data gets stored is different for these cases.
I also added a new selector pen=0123.. to choose which colours you want to have in your sprite. Example: I want a 4 colour sprite pen=2,3,4,5 for example. Then sp65 creates a BPP_2 sprite instead of how many colours are in the pcx file.
Later I also fixed the sp65 to understand the new storage format.

**Comment by Fabrizio-Caruso** [comment]
@karrika Sorry to bother you again. I am not able to compile a very simple example. I must be doing something wrong and stupid:
```
#include <tgi.h>
#include <6502.h>
#include <lynx.h>
#include <time.h>

const unsigned char freeze[] = {
    0x03,0x10,0x00,
    0x03,0x54,0x00,
    0x03,0x38,0x00,
    0x03,0xEE,0x00,
    0x03,0x38,0x00,
    0x03,0x54,0x00,
    0x03,0x10,0x00,
    0x00,
};


SCB_REHV_PAL Sfreeze = {
        BPP_1 | TYPE_NORMAL,
        LITERAL | REHV,
        NO_COLLIDE,
        0,
        freeze,
        10, 10,
        0x100, 0x100,
        {0x04}
};

int main(void)
{
    tgi_install (tgi_static_stddrv);
    tgi_init();		
    CLI();
    while (tgi_busy()){};
    SCB_REHV_PAL *spr;
    spr = &Sfreeze;

    tgi_setpalette(tgi_getdefpalette());
    tgi_setcolor(COLOR_WHITE);
    tgi_bar(0,0,159,101);
    tgi_sprite(&Sfreeze);
    tgi_updatedisplay();   
    while(1){};
    return 0;
}
```
which I am trying to compile as a single file with `cl65 -t lynx lynx_test.c` which should compile and link but fails with:
```
lynx_test.c:28: Warning: Pointer conversion to 'unsigned char *' from 'const unsigned char * const' discards qualifiers
lynx_test.c:42: Error: Variable identifier expected
lynx_test.c:42: Error: Undefined symbol: 'spr'
lynx_test.c:43: Warning: Converting pointer to integer without a cast
2 errors and 2 warnings generated.
```

**Reply by karrika** [reply]
You can only declare new variables at the start of a block - not in the middle of it.

The warning is a bug in the compiler imho.
"const unsigned char * const" is not what I declared. The error message is wrong.

const unsigned char a[] = {0];

unsigned char *b;

b = a;

Should not give a warning claiming the pointer is const. Unless i declare it

const unsigned char * const b = a;

Which would be silly as b could not be changed after that.

**Reply by Fabrizio-Caruso** [reply]
Thanks! Sorry, I should have figured it out myself.

**Comment by Fabrizio-Caruso** [comment]
I fear that displaying an 8x8 solid 2-color block (with no spill of its background) is harder than I thought because of the hardware bug (on the right most pixel?).

If I follow your suggestion I get an 8x16 block. I suppose it comes from the 0x00 values at the end of each line. Is there a solution? Using a different `hsize` produces a wrong image.
Maybe I could solve this in a very ugly way by displaying a totally empty 8x8 block and then a transparent "tile". Would it work? Is it doable? How? No better way? 

![image](https://user-images.githubusercontent.com/29057615/156810777-c059dbde-cafa-4dbd-8e25-fcdc0892e7c0.png)

```
#include <tgi.h>
#include <6502.h>
#include <lynx.h>
#include <time.h>

const unsigned char freeze[] = {
    0x03,0x10,0x00,
    0x03,0x54,0x00,
    0x03,0x38,0x00,
    0x03,0xEE,0x00,
    0x03,0x38,0x00,
    0x03,0x54,0x00,
    0x03,0x10,0x00,
    0x00,
};

SCB_REHV_PAL Sfreeze = {
        BPP_1 | TYPE_NORMAL,
        LITERAL | REHV,
        NO_COLLIDE,
        0,
        freeze,
        10, 10,
        0x100, 0x100,
        {0x14}
};

int main(void)
{
    SCB_REHV_PAL *spr;
    spr = &Sfreeze;
    
    tgi_install (tgi_static_stddrv);
    tgi_init();		
    CLI();
    while (tgi_busy()){};
    tgi_setpalette(tgi_getdefpalette());
    tgi_setcolor(COLOR_WHITE);
    tgi_bar(0,0,159,101);
    tgi_sprite(&Sfreeze);
    tgi_updatedisplay();
    
    while(1){};
    return 0;
}
```

**Reply by karrika** [reply]
You should be able to make a 8x8 fille block like:
/*
 * This file was generated by sp65 2.13.9 from
 * Slice of tiles.pcx (8x8, 4 colors, indexed)
 */

#define block_BPP          1
#define block_WIDTH        8
#define block_HEIGHT       8
const unsigned char block[] = {
    0x03,0xFF,0x00,0x03,0xFF,0x00,0x03,0xFF,
    0x00,0x03,0xFF,0x00,0x03,0xFF,0x00,0x03,
    0xFF,0x00,0x03,0xFF,0x00,0x03,0xFF,0x00,
    0x00,
};

#define block_BPP          1
#define block_WIDTH        16
#define block_HEIGHT       8
const unsigned char block[] = {
    0x04,0xFF,0xFF,0x00,0x04,0xFF,0xFF,0x00,
    0x04,0xFF,0xFF,0x00,0x04,0xFF,0xFF,0x00,
    0x04,0xFF,0xFF,0x00,0x04,0xFF,0xFF,0x00,
    0x04,0xFF,0xFF,0x00,0x04,0xFF,0xFF,0x00,
    0x00,
};

It appears that the Lynx does not do what it should with these blocks. Some "special" case again...

**Reply by karrika** [reply]
I did try with a sprite like this:

const unsigned char block[] = {
    0x02,0xAA,
    0x02,0x55,
    0x02,0x00,
    0x02,0xFF,
    0x03,0xAA,0xAA,
    0x03,0x55,0x55,
    0x03,0x00,0x00,
    0x03,0xFF,0xFF,
    0x00,
};

It produces exactly what was expected:
![Screenshot from 2022-03-05 09-43-55](https://user-images.githubusercontent.com/6996727/156873964-3a894a63-073c-4821-924e-c27d611843a0.png)

Buggy stuff again. One pixel is missing at the end of the bytes when the last pixel is one.

**Reply by karrika** [reply]
Now with extra 0x00 only when the last pixel is one.

const unsigned char block[] = {
    0x02,0xAA,
    0x03,0x55,0x00,
    0x02,0x00,
    0x03,0xFF,0x00,
    0x03,0xAA,0xAA,
    0x04,0x55,0x55,0x00,
    0x03,0x00,0x00,
    0x04,0xFF,0xFF,0x00,
    0x00,
};

And success, I hope...
![Screenshot from 2022-03-05 09-53-51](https://user-images.githubusercontent.com/6996727/156874279-318b74ce-abd6-4dbb-86fe-77423bc8c381.png)

So my previous comment that the bug does not occur in literal sprites was wrong. It affects all the sprites.

**Reply by karrika** [reply]
But this does not seem to be the whole truth.
![Screenshot from 2022-03-05 09-57-35](https://user-images.githubusercontent.com/6996727/156874415-a52c6f91-38bf-4256-92f6-b975e29c283c.png)

When I try to use the generated sprite with 2 colours instead of with transparency we have problems again.

**Reply by karrika** [reply]
I would skip the BPP_1 sprites and use more bits per pixel to get rid of these bugs. Perhaps a BPP_2 sprite. If you keep the lowest bit always zero then you never hit the bug. And you never add the extra 0 byte.
00 - colour 0
01 - not used
10 - colour 1
11 - not used

And... Success!
![Screenshot from 2022-03-05 10-12-51](https://user-images.githubusercontent.com/6996727/156874893-a66718a3-bf7c-4f16-a00f-1077db209807.png)

const unsigned char block[] = {
    0x03,0xAA,0xAA,
    0x03,0xAA,0xAA,
    0x03,0xAA,0xAA,
    0x03,0xAA,0xAA,
    0x03,0xAA,0xAA,
    0x03,0xAA,0xAA,
    0x03,0xAA,0xAA,
    0x03,0xAA,0xAA,
    0x00,
};

**Reply by Fabrizio-Caruso** [reply]
In my very specific case, I can assume the background to be the same for all "tiles" and equal to the background of the screen.
So BBP_1 seems to work because the bug only affects the background color of monochromatic "tiles". If the background color is set to the screen background, I do not see any bug but maybe I have not tested it well enough.

**Reply by Fabrizio-Caruso** [reply]
@karrika Indeed, this means I will have to map each byte to two bytes that replace 0 with 00 and 1 with 10, e.g.,
%01 00 11 10 -> (%00 10 00 00, %10 10 10 00).
My current solution uses BBP1 and forces me to delete the left column of pixels before drawing which produces horrible random flashing. I have not found a wait to wait for vsync (in single buffer mode) to hide this clean up.
But thanks to you I already got something working, which is very satisfying nevertheless.
![XChase_AtariLynx](https://user-images.githubusercontent.com/29057615/158013524-031f5054-ec04-4134-bb38-61eb72afb18a.jpg)
I will try to implement a BBP2 solution that does not require the ugly clean up of the last column of pixels.

**Reply by greg-king5** [reply]
> My current solution uses BBP1, and forces me to delete the right column of pixels before drawing which produces horrible random flashing. I have not found a way to wait for vsync (in single buffer mode) to hide this clean up.

You can use the page-flipping functions to sync; but then, put the driver back into single-buffer mode:
```c
tgi_updatedisplay();
while (tgi_busy()) {}
tgi_setdrawpage(0);  // Both view, and draw on, page zero.
```

**Reply by Fabrizio-Caruso** [reply]
@greg-king5 if I do as you say, I get a flashing screen. 
I am working in single buffer mode. Because of the bug on the last column of pixels, I need to delete that last column before I draw it as the background color will always be transparent (because of the bug). Ideally even in single buffer mode, there should be a way to wait for vsync and do the deletion when the screen is not visible but I have not found a way to do it.

**Reply by Fabrizio-Caruso** [reply]
@greg-king5 This is my ugly solution:
```
void _tgi_gfx_clean_bug(uint8_t x, uint8_t y)
{
    clean_bug_spr->hpos = (x)*8;
    clean_bug_spr->vpos = (y)*6;
    tgi_sprite(clean_bug_spr);
}


void _tgi_gfx_draw(uint8_t x, uint8_t y, uint8_t tile, uint8_t color)
{
    clean_bug(x,y);
    _tgi_tile[tile].hpos = (x)*8;
    _tgi_tile[tile].vpos = (y)*6;
    _tgi_tile[tile].penpal[0]=(unsigned char) (0x11+(color));
    tgi_sprite(&(_tgi_tile[tile]));
}
```

The good solution is what @karrika suggests, i.e., using BBP2 to avoid a final 1 bit.

**Reply by karrika** [reply]
There is also a technique where painting the whole screen takes ages and you want to draw just one moving sprite on top of it.

You set the drawpage to 1 and draw the complicated page. After that you use memcpy to copy the page 1 to page 0 and draw your sprite on it. I used this on solitaire card games where it was not possible to fit enough sprites in memory.

```
void bgcopy()
{
    asm("sei");
    asm("lda #$0f");
    asm("sta $fff9");
    memcpy(0xc038, 0xe018, 8160);
    asm("lda #$0c");
    asm("sta $fff9");
    asm("cli");
}

```

So it is a single buffer application with a store for the background canvas.

**Comment by Fabrizio-Caruso** [comment]
@karrika Sorry to bother you again. I hope I am not abusing your support.
I probably lack some basic understand of the Lynx hardware and TGI apis.
I am getting a flashing background screen and some missing "tiles" when I use `tgi_updatedisplay()`:
```c
#include <tgi.h>
#include <6502.h>
#include <lynx.h>
#include <time.h>

unsigned char freeze[] = {
    0x02,0x10,
    0x02,0x54,
    0x02,0x38,
    0x02,0xEE,
    0x02,0x38,
    0x02,0x54,
    0x02,0x10,
    0x00,
};

SCB_REHV_PAL Sfreeze = {
        BPP_1 | TYPE_NORMAL,
        LITERAL | REHV,
        NO_COLLIDE,
        0,
        freeze,
        10, 10,
        0x100, 0x100,
        {0x14}
};

int main(void) {
    unsigned char i;
    unsigned int j;
    
    SCB_REHV_PAL *spr;
    spr = &Sfreeze;
    
    tgi_install (tgi_static_stddrv);
    tgi_init();		
    CLI();
    while (tgi_busy()){};

    tgi_setpalette(tgi_getdefpalette());
    tgi_setcolor(COLOR_WHITE);
    tgi_bar(0,0,159,101);
    
    for(i=0;i<10;++i) {
        while (tgi_busy()){};
        Sfreeze.hpos += 8;
        tgi_sprite(&Sfreeze);
        tgi_updatedisplay();
        for(j=0;j<10000;++j){};
    }
    
    while(1){};
    return 0;
}
```

**Reply by karrika** [reply]
Obviously.

You need to paint the background during every display cycle. Something like:
```c
while (1) {
    while (tgi_busy())
        ;
    // Now we are drawing on the draw page - not on screen
    tgi_setcolor(COLOR_WHITE);
    tgi_bar(0, 0, 159, 101);
    tgi_sprite(&Sfreeze);
    // Now we want to swap the draw page with the display page
    tgi_updatedisplay();
    // The Lynx system will now wait for the VBL interrupt.
    // Inside the interrupt routine the swap will take place.
    // This allows for flicker free displays
    // The tgi_busy() returns true until the swap has occurred
}
```

**Reply by Fabrizio-Caruso** [reply]
Ok, so I need to redraw everything after each change on the screen. The screen has no memory. I cannot just draw on the screen with `tgi_sprite`.
It means I can only do this with a buffer. I need to have an array to store all tiles position and do a single `tgi_updatedisplay` at the end of my game main loop.

**Reply by Fabrizio-Caruso** [reply]
If I remove `tgi_updatedisplay`, I still see the sprites. So `tgi_sprite` does produce a result on the display page even without `tgi_updatedisplay`.
This code seems to work:
```c
#include <tgi.h>
#include <6502.h>
#include <lynx.h>
#include <time.h>

unsigned char freeze[] = {
    0x02,0x10,
    0x02,0x54,
    0x02,0x38,
    0x02,0xEE,
    0x02,0x38,
    0x02,0x54,
    0x02,0x10,
    0x00,
};

SCB_REHV_PAL Sfreeze = {
        BPP_1 | TYPE_NORMAL,
        LITERAL | REHV,
        NO_COLLIDE,
        0,
        freeze,
        10, 10,
        0x100, 0x100,
        {0x14}
};

int main(void)
{
    unsigned char i;
    unsigned int j;
    
    SCB_REHV_PAL *spr;
    spr = &Sfreeze;
    
    tgi_install (tgi_static_stddrv);
    tgi_init();		
    CLI();
    while (tgi_busy()){};

    tgi_setpalette(tgi_getdefpalette());
    tgi_setcolor(COLOR_BLACK);
    tgi_bar(0,0,159,101);
    
    for(i=0;i<10;++i)
    {
        Sfreeze.hpos += 8;
        tgi_sprite(&Sfreeze);
        for(j=0;j<10000;++j){};
    }
    
    while(1){};
    return 0;
}
```

**Reply by Fabrizio-Caruso** [reply]
@karrika it seems I am writing directly onto the display buffer, which is what I want to do as I don't mind about subtle effects such as tearing (not for the moment at least).
Is this the normal behavior on real hardware? I am testing this with the Handy emulator.

**Reply by greg-king5** [reply]
The driver has two modes: "single-bufferred" and "double-bufferred, page-flipping".  The TGI kernel sets it to single-bufferred mode.  (It also sets the palette, you don't need to do that.)  The first call to `tgi_updatedisplay()` changes the driver into page-flipping mode.  But, you don't care about page-flipping, yet.  `tgi_busy()` is used _after_ `tgi_updatedisplay()`.  It's pointless to use it at the beginning of a program.

Also, don't use `CLI()`, the start-up code enables interrupts.

**Reply by karrika** [reply]
You don't need tgi_busy() at all in the single-buffered mode.

**Reply by polluks** [reply]
@karrika Maybe you can add a comment to the [example](https://cc65.github.io/doc/tgi.html#ss1.28).
